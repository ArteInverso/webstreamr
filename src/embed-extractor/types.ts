import { StreamWithMeta } from '../types';

export interface EmbedExtractor {
  readonly id: string;

  readonly label: string;

  readonly extract: (embedUrl: string, language: string) => Promise<StreamWithMeta>;
}
