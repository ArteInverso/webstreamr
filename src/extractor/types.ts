import { Context, CountryCode, UrlResult } from '../types';

export interface Extractor {
  readonly id: string;

  readonly label: string;

  readonly ttl: number;

  readonly supports: (ctx: Context, url: URL) => boolean;

  readonly normalize: (url: URL) => URL;

  readonly extract: (ctx: Context, url: URL, countryCode: CountryCode, title?: string | undefined) => Promise<UrlResult[]>;
}
