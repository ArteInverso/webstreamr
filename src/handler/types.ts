import { ContentType } from 'stremio-addon-sdk';
import { Context, CountryCode } from '../types';
import { Id } from '../utils';

export interface HandleResult {
  countryCode: CountryCode;
  referer?: URL;
  title?: string;
  url: URL;
}

export interface Handler {
  readonly id: string;

  readonly label: string;

  readonly contentTypes: ContentType[];

  readonly countryCodes: CountryCode[];

  readonly handle: (ctx: Context, type: ContentType, id: Id) => Promise<(HandleResult[])>;
}
