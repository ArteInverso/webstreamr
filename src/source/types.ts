import { ContentType } from 'stremio-addon-sdk';
import { Context, CountryCode } from '../types';
import { Id } from '../utils';

export interface SourceResult {
  countryCode: CountryCode;
  title?: string;
  url: URL;
}

export interface Source {
  readonly id: string;

  readonly label: string;

  readonly contentTypes: ContentType[];

  readonly countryCodes: CountryCode[];

  handle(ctx: Context, type: ContentType, id: Id): Promise<(SourceResult[])>;
}
