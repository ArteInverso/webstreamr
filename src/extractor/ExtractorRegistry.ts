import TTLCache from '@isaacs/ttlcache';
import winston from 'winston';
import { Extractor } from './types';
import { Context, UrlResult } from '../types';
import { Fetcher } from '../utils';
import { DoodStream } from './DoodStream';
import { Dropload } from './Dropload';
import { SuperVideo } from './SuperVideo';
import { ExternalUrl } from './ExternalUrl';
import { NotFoundError } from '../error';

export class ExtractorRegistry {
  private readonly logger: winston.Logger;
  private readonly extractors: Extractor[];
  private readonly urlResultCache: TTLCache<string, UrlResult | undefined>;

  constructor(logger: winston.Logger, fetcher: Fetcher) {
    this.logger = logger;
    this.extractors = [
      new DoodStream(fetcher),
      new Dropload(fetcher),
      new SuperVideo(fetcher),
      new ExternalUrl(fetcher), // fallback extractor which must come last
    ];
    this.urlResultCache = new TTLCache({ max: 1024 });
  }

  readonly handle = async (ctx: Context, url: URL, countryCode: string): Promise<UrlResult | undefined> => {
    let urlResult = this.urlResultCache.get(url.href);
    if (this.urlResultCache.has(url.href)) {
      return urlResult;
    }

    const extractor = this.extractors.find(extractor => extractor.supports(url)) as Extractor;
    this.logger.info(`Extract stream URL using ${extractor.id} extractor from ${url}`, ctx);

    try {
      urlResult = await extractor.extract(ctx, url, countryCode);
    } catch (error) {
      if (error instanceof NotFoundError) {
        this.urlResultCache.set(url.href, urlResult, { ttl: extractor.ttl });

        return undefined;
      }

      this.logger.warn(`${extractor.id} error: ${error}`, ctx);
      return undefined;
    }

    this.urlResultCache.set(url.href, urlResult, { ttl: extractor.ttl });

    return urlResult;
  };
}
