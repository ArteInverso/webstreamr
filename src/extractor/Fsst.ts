import * as cheerio from 'cheerio';
import { Extractor } from './types';
import { Fetcher } from '../utils';
import { Context, CountryCode, UrlResult } from '../types';

export class Fsst implements Extractor {
  readonly id = 'fsst';

  readonly label = 'Fsst';

  readonly ttl = 900000; // 15m

  private readonly fetcher: Fetcher;

  constructor(fetcher: Fetcher) {
    this.fetcher = fetcher;
  }

  readonly supports = (_ctx: Context, url: URL): boolean => null !== url.host.match(/fsst/);

  readonly normalize = (url: URL): URL => url;

  readonly extract = async (ctx: Context, url: URL, countryCode: CountryCode): Promise<UrlResult[]> => {
    const html = await this.fetcher.text(ctx, url);

    const $ = cheerio.load(html);
    const title = $('title').text().trim();

    const filesMatch = html.match(/file:"(.*)"/) as string[];

    return (filesMatch[1] as string).split(',').map((fileString, index) => {
      const heightAndUrlMatch = fileString.match(/\[([\d]+)p\](.*)/) as string[];

      return {
        url: new URL(heightAndUrlMatch[2] as string),
        label: this.label,
        sourceId: `${this.id}_${countryCode}_${index}`,
        meta: {
          countryCode,
          height: parseInt(heightAndUrlMatch[1] as string),
          title,
        },
      };
    });
  };
}
