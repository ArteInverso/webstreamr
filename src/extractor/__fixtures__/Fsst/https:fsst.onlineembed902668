<html>
	<head>
		<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
		<title>How To Train Your Dragon 2025.mp4</title>
					<script src='/player/player-101224.js' type='text/javascript'></script>
				<!-- Google tag (gtag.js) -->
	<script async src="https://www.googletagmanager.com/gtag/js?id=G-CFYS4N19SL"></script>
	<script>
	  window.dataLayer = window.dataLayer || [];
	  function gtag(){dataLayer.push(arguments);}
	  gtag('js', new Date());
	
	  gtag('config', 'G-CFYS4N19SL', {cookie_flags: 'max-age=7200;secure;samesite=none'});
	  gtag('config', 'G-SZ2D21EHGZ', {cookie_flags: 'max-age=7200;secure;samesite=none'});
	  gtag('config', 'G-7F9RVQ2TM0', {cookie_flags: 'max-age=7200;secure;samesite=none'});
	  gtag('config', 'G-LZP7HENPSC', {cookie_flags: 'max-age=7200;secure;samesite=none'});
	</script>
	</head>
			<body style="margin:0; padding:0; background:#000000; color:#ffffff">
	<div id="Video902668" style="width:100%; height:100%;"></div>
																																						
			
	
			

	
					

		
		
	<script>
		var isMobile = {
			Android: function() {
				return navigator.userAgent.match(/Android/i);
			},
			BlackBerry: function() {
				return navigator.userAgent.match(/BlackBerry/i);
			},
			iOS: function() {
				return navigator.userAgent.match(/iPhone|iPad|iPod/i);
			},
			Opera: function() {
				return navigator.userAgent.match(/Opera Mini/i);
			},
			SmartTV: function() {
				return navigator.userAgent.match(/Apple TV|Tizen|Web0S|WebOS|TSBNetTV|PHILIPSTV|NETFLIX|BRAVIA|Opera TV Store|SmartTV|SmartTv/i);
			},
			Windows: function() {
				return navigator.userAgent.match(/IEMobile/i);
			},
			any: function() {
				return (isMobile.Android() || isMobile.BlackBerry() || isMobile.iOS() || isMobile.Opera() || isMobile.SmartTV() || isMobile.Windows());
			}
		};
		if(isMobile.iOS()){
			var player = new Playerjs({ 
				id: 'Video902668', 
								subtitle: "",
				poster: 'https://www.secvideo1.online/contents/videos_screenshots/902000/902668/preview.jpg',
				file:"https://www.secvideo1.online/get_file/10/648fc9b266670e9ad6756a066a68a8db0b36868457/902000/902668/902668_360p.mp4/",
												url: 'https://csst.online/videos/902668/how-to-train-your-dragon-2025-mp4/',
				embed: "https://csst.online/embed/902668/",
				default_quality: "360p",
				preroll: "//4736.in/vpaid/?id=906",			});
		} else {
		if(isMobile.Android()){
			var player = new Playerjs({ 
				id: 'Video902668', 
								subtitle: "",
				poster: 'https://www.secvideo1.online/contents/videos_screenshots/902000/902668/preview.jpg',
				file:"https://www.secvideo1.online/get_file/10/648fc9b266670e9ad6756a066a68a8db0b36868457/902000/902668/902668_360p.mp4/",
												url: 'https://csst.online/videos/902668/how-to-train-your-dragon-2025-mp4/',
				embed: "https://csst.online/embed/902668/",
				default_quality: "360p",
				preroll: "//4736.in/vpaid/?id=906",			});
		} else {
		if(isMobile.SmartTV()){
			var player = new Playerjs({ 
				id: 'Video902668', 
								subtitle: "",
				poster: 'https://www.secvideo1.online/contents/videos_screenshots/902000/902668/preview.jpg',
				file:"https://www.secvideo1.online/get_file/10/648fc9b266670e9ad6756a066a68a8db0b36868457/902000/902668/902668_360p.mp4/",
												url: 'https://csst.online/videos/902668/how-to-train-your-dragon-2025-mp4/',
				embed: "https://csst.online/embed/902668/",
				default_quality: "360p",
				preroll: "//4736.in/vpaid/?id=906",			});
		} else {
			var player = new Playerjs({ 
				id: 'Video902668', 
								subtitle: "",
				poster: 'https://www.secvideo1.online/contents/videos_screenshots/902000/902668/preview.jpg',
				file:"https://www.secvideo1.online/get_file/10/648fc9b266670e9ad6756a066a68a8db0b36868457/902000/902668/902668_360p.mp4/",
												url: 'https://csst.online/videos/902668/how-to-train-your-dragon-2025-mp4/',
				embed: "https://csst.online/embed/902668/",
				default_quality: "360p",
				preroll: "//4736.in/vpaid/?id=906",			});
		}
		}
		}
	</script>  
	<script>
	function PlayerjsEvents(event,id,info){
	   if(event=="vast_url"){
			url = info; 
		}
	   if(event=="inited"){
		  console.log("inited");
		  console.log(Math.round(player.api("time")));
		  window.parent.postMessage({
			event: 'inited',
			time: Math.round(player.api("time")),
		}, '*');
	   }
	   if(event=="vast_Impression"){
		  console.log("Impression");
		  console.log(url);
		  yaCounter49597111.reachGoal('vast_impression', {Ads: url.split(/\/+/)[1]});
		  gtag('event', 'ADS_viewed', {
						  'send_to': 'G-LZP7HENPSC',
						  'event_category': 'Video',
						  'event_label': '28214',
						});
		  gtag('event', 'Vast Impression', {
						  'send_to': 'G-7F9RVQ2TM0',
						  'event_category': 'Video',
						  'event_label': url.split(/\/+/)[1],
						});
		}
	   if(event=="start"){
		  console.log("started");
		  console.log(Math.round(player.api("time")));
		  yaCounter49597111.reachGoal('play');
		  gtag('event', 'Video_played', {
						  'send_to': 'G-LZP7HENPSC',
						  'event_category': 'Video',
						  'event_label': '28214',
						});
			gtag('event', 'Play', {
						  'send_to': 'G-SZ2D21EHGZ',
						  'event_category': 'Video',
						  'event_label': 'Play',
						});
		  window.parent.postMessage({
			event: 'started',
			time: Math.round(player.api("time")),  
		}, '*');
	   }
	   if(event=="pause"){
		  console.log("paused");
		  console.log(Math.round(player.api("time")));
		  window.parent.postMessage({
			event: 'paused',
			time: Math.round(player.api("time")),
		}, '*');
	   }
	   if(event=="end"){
		  console.log("ended");
		  console.log(Math.round(player.api("time")));
		  window.parent.postMessage({
			event: 'ended',
			time: Math.round(player.api("time")),
		}, '*');  
	   }
	}
	</script>
<!-- Yandex.Metrika counter -->
<script type="text/javascript" >
    (function (d, w, c) {
        (w[c] = w[c] || []).push(function() {
            try {
                w.yaCounter49597111 = new Ya.Metrika2({
                    id:49597111,
                    clickmap:true,
                    trackLinks:true,
                    accurateTrackBounce:true
                });
            } catch(e) { }
        });

        var n = d.getElementsByTagName("script")[0],
            s = d.createElement("script"),
            f = function () { n.parentNode.insertBefore(s, n); };
        s.type = "text/javascript";
        s.async = true;
        s.src = "https://cdn.jsdelivr.net/npm/yandex-metrica-watch/tag.js";

        if (w.opera == "[object Opera]") {
            d.addEventListener("DOMContentLoaded", f, false);
        } else { f(); }
    })(document, window, "yandex_metrika_callbacks2");
</script>
<noscript><div><img src="https://mc.yandex.ru/watch/49597111" style="position:absolute; left:-9999px;" alt="" /></div></noscript>
<!-- /Yandex.Metrika counter -->
<!--  -->
<!-- 2 -->
<!--  -->

	</body>
</html>