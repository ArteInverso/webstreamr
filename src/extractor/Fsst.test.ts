import winston from 'winston';
import { FetcherMock } from '../utils';
import { CountryCode } from '../types';
import { ExtractorRegistry } from './ExtractorRegistry';
import { Fsst } from './Fsst';
import { createTestContext } from '../test';

const logger = winston.createLogger({ transports: [new winston.transports.Console({ level: 'nope' })] });
const extractorRegistry = new ExtractorRegistry(logger, [new Fsst(new FetcherMock(`${__dirname}/__fixtures__/Fsst`))]);

const ctx = createTestContext();

describe('Fsst', () => {
  test('Blood & Sinners', async () => {
    expect(await extractorRegistry.handle(ctx, new URL('https://fsst.online/embed/900576/'), CountryCode.de)).toMatchSnapshot();
  });

  test('Dead City', async () => {
    expect(await extractorRegistry.handle(ctx, new URL('https://fsst.online/embed/901994/'), CountryCode.de)).toMatchSnapshot();
  });

  test('How to Train Your Dragon', async () => {
    expect(await extractorRegistry.handle(ctx, new URL('https://fsst.online/embed/902668/'), CountryCode.de)).toMatchSnapshot();
  });
});
