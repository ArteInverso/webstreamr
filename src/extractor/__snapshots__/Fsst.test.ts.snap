// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Fsst Blood & Sinners 1`] = `
[
  {
    "label": "Fsst",
    "meta": {
      "countryCode": "de",
      "height": 360,
      "title": "Blood.and.Sinners.2025.mkv",
    },
    "sourceId": "fsst_de_0",
    "url": "https://www.secvideo1.online/get_file/16/a135ab46b1243e41c0282c16d7cdcbcd3e54de0f28/900000/900576/900576_360p.mp4/",
  },
  {
    "label": "Fsst",
    "meta": {
      "countryCode": "de",
      "height": 720,
      "title": "Blood.and.Sinners.2025.mkv",
    },
    "sourceId": "fsst_de_1",
    "url": "https://www.secvideo1.online/get_file/16/801a42b7ac11f7e08101fe8c3cb5370470f5f91590/900000/900576/900576_720p.mp4/",
  },
  {
    "label": "Fsst",
    "meta": {
      "countryCode": "de",
      "height": 1080,
      "title": "Blood.and.Sinners.2025.mkv",
    },
    "sourceId": "fsst_de_2",
    "url": "https://www.secvideo1.online/get_file/16/fae180e05d8d7af50fff44c7a24232b522db464935/900000/900576/900576.mp4/",
  },
]
`;

exports[`Fsst Dead City 1`] = `
[
  {
    "label": "Fsst",
    "meta": {
      "countryCode": "de",
      "height": 360,
      "title": "the.walking.dead.dead.city.s02e06.german.dl.1080p.web.x264-wvf.mkv",
    },
    "sourceId": "fsst_de_0",
    "url": "https://www.secvideo1.online/get_file/9/99510c9abde68b5e82b08cd239172405b06f84c9e5/901000/901994/901994_360p.mp4/",
  },
  {
    "label": "Fsst",
    "meta": {
      "countryCode": "de",
      "height": 720,
      "title": "the.walking.dead.dead.city.s02e06.german.dl.1080p.web.x264-wvf.mkv",
    },
    "sourceId": "fsst_de_1",
    "url": "https://www.secvideo1.online/get_file/9/868ae0229e543e7a9a6f7cf4b3df88f460a1bfc444/901000/901994/901994_720p.mp4/",
  },
  {
    "label": "Fsst",
    "meta": {
      "countryCode": "de",
      "height": 1080,
      "title": "the.walking.dead.dead.city.s02e06.german.dl.1080p.web.x264-wvf.mkv",
    },
    "sourceId": "fsst_de_2",
    "url": "https://www.secvideo1.online/get_file/9/44ff1bf6024db32840054f00029afb114120b91353/901000/901994/901994.mp4/",
  },
]
`;
